File: Algorithms/C++/01768-Merge Strings Alternately.cpp
class Solution {
public:
    string mergeAlternately(string word1, string word2) {
        int p1{0};
        int p2{0};
        std::vector<char> res;
        while(p1 < word1.size() || p2 < word2.size()){
            if(p1 < word1.size()) res.push_back(word1[p1]);
            if(p2 < word2.size()) res.push_back(word2[p2]);
            p1++;
            p2++;
        }

        // change result char array into string
        return std::string(res.begin(), res.end());
    }
};

/*
    2 pointers
*/

File: Algorithms/Python3/01768-Merge Strings Alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        minLen = min(len(word1), len(word2))

        merged = ""
        for i in range(minLen):
            merged += word1[i]
            merged += word2[i]

        len1 = len(word1)
        len2 = len(word2)

        if len1 == len2:
            return merged
        elif len1 > len2:
            merged += word1[minLen:len1]
        else:
            merged += word2[minLen:len2]
        
        return merged


'''
    2 pointer 
        - find the minimum length of both words (minLen)
        - iterate through minLen times for both words
        - figure out the larger word
        - append the rest of the larger word to the end of output
        - return output

    tc: O(n)
    sc: O(1)


'''